packages/grid/x-grid-modules/src/components/checkbox-renderer.tsx

```diff
@@ -4,9 +4,14 @@ import { SelectionChangeParams } from '../models/params/selectionChangeParams';
 import { ColParams } from '../models/params/colParams';
 import { CellParams } from '../models/params/cellParams';
 
-export const HeaderCheckbox: React.FC<ColParams> = React.memo(({ api }) => {
-  const [isChecked, setChecked] = React.useState(false);
-  const [isIndeterminate, setIndeterminate] = React.useState(false);
+export const HeaderCheckbox: React.FC<ColParams> = ({ api }) => {
+  const selectedRows = api.getSelectedRows();
+  const allRowsCount = api.getAllRowIds().length;
+  const isCurrentIndeterminate = selectedRows.length > 0 && selectedRows.length !== selectedRows;
+  const isCurrentChecked = selectedRows.length === allRowsCount || isCurrentIndeterminate;
+
+  const [isChecked, setChecked] = React.useState(isCurrentChecked);
+  const [isIndeterminate, setIndeterminate] = React.useState(isCurrentIndeterminate);
```

Comment:
Does `api` not change if we select a different row? Seems very brittle to have individual components duplicate the logic of which row is considered checked and which isn't.

Reply:
this is just the header checkbox

Reply:
It always starts out with "just this small component". If we have the choice between robust and brittle we might as well use robust. I guess `api` is not changing if the underlying data changes? How are "subscribers" usually notified when the data changes?

Reply:
No Api is not changing. They can use an event handler to get latest data.